### PR TITLE
DOC Add troubleshooting section for hybrid architecture target_modules

### LINKED ISSUE #3554: NemotronH LoRA: default target_modules only cover 7% of layers, silent failure
## Problem

PEFT's default LoRA target modules for `nemotron_h` (added in PR #3289) only target attention projections (`q_proj`, `k_proj`, `v_proj`, `o_proj`). For NemotronH's hybrid Mamba-2 + Attention architecture, this means LoRA only touches **4 out of 56 layers (7.1%)** — the other 52 layers are Mamba SSM layers with different parameter names (`in_proj`, `out_proj`, `A_log`, `D`, `dt`, `x_proj`, `z_proj`).

Worse, PEFT **silently skips** target module names that don't exist in the model. There's no warning that the configured target modules only matched a small fraction of layers. The user sees training "complete" with loss not improving, and has no idea why.

## Reproduction

We ran DPO (LoRA) on `nvidia/NVIDIA-Nemotron-Nano-9B-v2` with the default PEFT target modules:

```python
from peft import LoraConfig
config = LoraConfig(
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],  # PEFT default for nemotron_h
    r=16, lora_alpha=32, lora_dropout=0.1,
)
```

Results:
- **Training completed** (25/25 steps, 12 minutes)
- **Loss stayed at 0.6931 (ln(2))** for all 25 steps — zero learning
- **HumanEval: 0.0% (0/164)** — the LoRA adapter produced no behavioral change

The Nemotron-Nano-9B-v2 has 56 layers with this hybrid pattern:
```
M-M-M-MM-M-M-M*-M-M-M*-M-M-M-M*-M-M-M-M*-M-MM-M-M-M-M-M-
```
Where `M` = Mamba, `*` = Attention. Only 4 layers are attention. The other 52 are Mamba layers with parameters `in_proj`, `out_proj`, `A_log`, `D` — none of which were targeted by LoRA.

## Root Cause

1. PEFT PR #3289 added `nemotron_h` to `TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING` with only attention modules
2. PEFT PR #2562 added `nemotron_h` to `mamba_model_types` in `_check_lora_target_modules_mamba()`, which **forbids** targeting Mamba modules (`in_proj`, `out_proj`, `conv1d`)
3. The combination means: PEFT defaults only target attention layers, and PEFT actively blocks you from targeting Mamba layers
4. No warning is emitted when target modules match an unusually low number of parameters

### Fix 1: Warn on low coverage

When `LoraConfig.target_modules` matches fewer than N% of linear layers in the model, emit a warning:

```python
if matched_layers / total_linear_layers < 0.3:
    warnings.warn(
        f"LoRA target modules only matched {matched_layers}/{total_linear_layers} "
        f"linear layers ({100*matched/total:.1f}%). For hybrid architectures like "
        f"NemotronH, consider including Mamba modules (in_proj, out_proj) in target_modules."
    )
```

### Fix 2: Document hybrid architecture target modules

Add documentation for hybrid Mamba+Attention models showing the correct target modules:

```python
# For NemotronH / Jamba / Granite-H hybrid models:
LoraConfig(
    target_modules=[
        # Attention layers
        "q_proj", "k_proj", "v_proj", "o_proj",
        # Mamba layers (requires removing from mamba_model_types check)
        "in_proj", "out_proj", "x_proj", "dt_proj",
    ],
    # Skip: conv1d, A_log, D (structured SSM parameters, not suitable for LoRA)
)
```

### Fix 3: Reconsider the Mamba module ban

PR #2562 banned Mamba modules from LoRA targeting due to shape issues. But `in_proj` and `out_proj` are standard `nn.Linear` layers — they should be LoRA-compatible. The ban should be scoped to `conv1d` and structured SSM parameters (`A_log`, `D`), not all Mamba modules.

## Environment

- peft: latest (with PR #3289 merged)
- transformers: 5.16.0.dev0
- Model: nvidia/NVIDIA-Nemotron-Nano-9B-v2 (56 layers, 4 attention + 52 Mamba)
- GPU: NVIDIA GB10 (119GB VRAM)

## Related

- PEFT PR #3289: Added default target_modules for nemotron_h (attention only)
- PEFT PR #2562: Added Mamba target module check (bans in_proj, out_proj)
- Unsloth issue #3854: "Trouble fine-tuning Nemotron 3 Nano, failure when merging LoRA"
- Full writeup: https://github.com/davidnichols-ops/claude-yolo-vibes-v5/blob/master/EVALUATION.md

### PR DESCRIPTION
**Pull request by davidnichols-ops**
Fixes #3554

## Summary

Adds a troubleshooting section for hybrid architectures (Mamba, Jamba, NemotronH) where `target_modules` entries that don't match any model layer are silently skipped, causing training to run with no learning.

## Why

As reported in #3554, users targeting only attention projections (`q_proj`, `k_proj`, `v_proj`, `o_proj`) on hybrid Mamba+Attention models such as `nemotron_h` can end up with only ~7% of layers trainable. PEFT raises an error only when *none* of the configured `target_modules` match; when some match and some don't, the non-matching names are silently skipped.

## What changed

- `docs/source/developer_guides/troubleshooting.md`: new subsection "Target modules don't match the model" under "Training runs but the model does not improve".

## Scope

Documentation only. No library code or tests changed.

## Related

- Issue: #3554
- Maintainer approval for docs fix: https://github.com/huggingface/peft/issues/3554#issuecomment-5326537060
- Closed, unmerged attempt with extra changes: #3600

## Checks run

- `python -m ruff check --fix src tests examples docs scripts docker` ✅
- `python -m ruff format src tests examples docs scripts docker` ✅
- `doc-builder style src/peft tests docs/source --max_len 119` ✅

## AI assistance

This PR was prepared with AI assistance and has been reviewed by the human submitter.

### CODE REVIEW

File: docs/source/developer_guides/troubleshooting.md [hunk 1]
```diff
@@ -114,6 +114,40 @@ To avoid this, use the following order:
...
+```
+
+If the reported number is much lower than expected, list the model's linear layers and compare them against your `target_modules`:
+
+```python
+import torch
+
+for name, module in model.named_modules():
+    if isinstance(module, torch.nn.Linear):
+        print(name)
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/troubleshooting.md]: I could actually be more indicative to print:

```python
print(model.get_layer_status())
```

Most users don't know about this feature, did you give it a try?

File: docs/source/developer_guides/troubleshooting.md [hunk 2]
```diff
@@ -114,6 +114,40 @@ To avoid this, use the following order:
...
+```
+
+Any `target_modules` entry that does not appear in this list is silently skipped. Update `target_modules` to match the actual module names in the model.
+
+> [!TIP]
+> This issue affects all PEFT methods that use `target_modules` (LoRA, LoHa, IA³, etc.), not just LoRA. The diagnostic steps are the same regardless of the method. If you are unsure which module names a given architecture uses, check the model's documentation or inspect `model.named_modules()` before configuring the adapter.
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/troubleshooting.md]: > inspect `model.named_modules()`

IMO, printing the model repr is generally more useful, especially since bigger models can have many hundreds of modules.

File: docs/source/developer_guides/troubleshooting.md [hunk 3]
```diff
@@ -114,6 +114,40 @@ To avoid this, use the following order:
...
+
+> [!TIP]
+> Some modules are forbidden by PEFT on certain architectures. For example, `out_proj` and `conv1d` are not allowed on Mamba-based models because the fused SSM kernels bypass adapter layers. PEFT raises a `ValueError` in these cases, so the error is visible — unlike the silent skipping described above.
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/troubleshooting.md]: Probably can be removed, as the user will get an indicative error when they try.

File: docs/source/developer_guides/troubleshooting.md [hunk 4]
```diff
@@ -114,6 +114,40 @@ To avoid this, use the following order:
 2. Wrap it with [`get_peft_model`].
 3. Only then register hooks or create the optimizer.
 
+## Target modules don't match the model
+
+If training runs without errors but the loss barely moves — or the model shows no improvement after fine-tuning — the configured `target_modules` may not match the actual module names in the model. This is common with hybrid architectures such as Mamba, Jamba, or NemotronH, which use different layer names than standard Transformer models.
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/troubleshooting.md]: > the configured `target_modules`

Worth a mention: or the default target modules if not specified otherwise by the user.

File: docs/source/developer_guides/troubleshooting.md [hunk 5]
```diff
@@ -114,6 +114,40 @@ To avoid this, use the following order:
 2. Wrap it with [`get_peft_model`].
 3. Only then register hooks or create the optimizer.
 
+## Target modules don't match the model
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/troubleshooting.md]: Placing this below the section `## Training runs but the model does not improve` is correct, but I think it should be a subsection.

```suggestion
### Target modules don't match the model
```

To make this fit with the section above, it should also introduce a subsection. I think a good place would be before the paragraph that starts with `If the weights did not move, a likely cause`. The subsection title could be:

```
### Missing forward or backward hooks
```

The text that comes before that with the small code snippet is pretty general and can stay under the `## Training runs but the model does not improve` parent section.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for adding this section to the troubleshooting docs. Generally it looks good, but I have a few suggestions, please check.

**Comment by davidnichols-ops** [comment]
Thanks for the review @BenjaminBossan! I've addressed all the suggestions in 0e688d3d:

1. **`get_layer_status()`** — Replaced the `named_modules()` loop with `print(model.get_layer_status())` and mentioned `print(model)` repr as an alternative for unfamiliar architectures.

2. **Default `target_modules`** — Added "(or the default `target_modules` if not specified otherwise by the user)" to the intro paragraph.

3. **Removed forbidden-modules TIP** — Dropped the second TIP about `out_proj`/`conv1d` since PEFT raises a `ValueError` in those cases (visible error, not silent skipping).

4. **Subsection hierarchy** — Changed `## Target modules don't match the model` to `###` under the parent `## Training runs but the model does not improve`, and added `### Missing forward or backward hooks` before the stale-hooks paragraph so both are subsections under the same parent.

`make style` passes (ruff check + format + doc-builder style).

**Review by BenjaminBossan — APPROVED** [review]
Thanks for improving the docs, LGTM.
